File: modules/volt-shield/src/VoltShieldModule.ts
import { NativeModule, requireNativeModule } from "expo";

export type SocialMediaUsageResult = {
  dailyAverageMinutes: number;
  todayMinutes: number;
  daysOfData: number;
};

export type ShieldRuntimeState = {
  enabled: boolean;
  unlockUntil: number;
};

declare class VoltShieldModule extends NativeModule {
  isUsageAccessGranted(): Promise<boolean>;
  openUsageAccessSettings(): Promise<void>;
  isAccessibilityServiceEnabled(): Promise<boolean>;
  openAccessibilitySettings(): Promise<void>;
  setBlockedPackages(packages: string[]): Promise<void>;
  getBlockedPackages(): Promise<string[]>;
  /** Sync Convex fuel bank to native (Android toast/notification + iOS Screen Time app group for shield placeholders). */
  setFuelMinutesAvailable(minutes: number): Promise<void>;
  setShieldEnabled(enabled: boolean): Promise<void>;
  setShieldUnlockUntil(timestampMs: number): Promise<void>;
  getShieldRuntimeState(): Promise<ShieldRuntimeState>;
  /** Query UsageStatsManager for social-media foreground time over the past N days. Android only. */
  getSocialMediaUsageMinutes(days: number): Promise<SocialMediaUsageResult>;
}

export default requireNativeModule<VoltShieldModule>("VoltShield");

File: modules/volt-shield/ios/VoltShieldModule.swift
import ExpoModulesCore

public class VoltShieldModule: Module {
  private let shieldEnabledKey = "voltShieldEnabled"
  private let shieldUnlockUntilKey = "voltShieldUnlockUntil"

  private func appGroupDefaults() -> UserDefaults? {
    guard
      let appGroup = Bundle.main.object(
        forInfoDictionaryKey: "REACT_NATIVE_DEVICE_ACTIVITY_APP_GROUP"
      ) as? String
    else {
      return nil
    }
    return UserDefaults(suiteName: appGroup)
  }

  public func definition() -> ModuleDefinition {
    Name("VoltShield")

    AsyncFunction("isUsageAccessGranted") {
      false
    }

    AsyncFunction("openUsageAccessSettings") {}

    AsyncFunction("isAccessibilityServiceEnabled") {
      false
    }

    AsyncFunction("openAccessibilitySettings") {}

    AsyncFunction("setBlockedPackages") { (_: [String]) in }

    AsyncFunction("getBlockedPackages") {
      [String]()
    }

    AsyncFunction("setFuelMinutesAvailable") { (minutes: Int) in
      guard let defaults = self.appGroupDefaults() else { return }
      defaults.set(minutes, forKey: "voltFuelMinutesAvailable")
    }

    AsyncFunction("setShieldEnabled") { (enabled: Bool) in
      self.appGroupDefaults()?.set(enabled, forKey: self.shieldEnabledKey)
    }

    AsyncFunction("setShieldUnlockUntil") { (timestampMs: Double) in
      self.appGroupDefaults()?.set(
        max(0, timestampMs),
        forKey: self.shieldUnlockUntilKey
      )
    }

    AsyncFunction("getShieldRuntimeState") {
      let defaults = self.appGroupDefaults()
      let enabled = defaults?.object(forKey: self.shieldEnabledKey) as? Bool ?? true
      let unlockUntil = defaults?.double(forKey: self.shieldUnlockUntilKey) ?? 0
      return [
        "enabled": enabled,
        "unlockUntil": unlockUntil,
      ] as [String: Any]
    }
  }
}
